src/hooks.rs

```diff
@@ -0,0 +1,196 @@
+            // doesn't provide us an API to pass argv directly. Instead, we have
+            // to seperate arg0 and the rest of the args.
+            let (arg0, args) = if let Some(mut args) = hook.args.clone() {
+                let arg0 = args.remove(0);
+                (arg0, args)
+            } else {
+                (hook.path.as_path().display().to_string(), vec![])
+            };
+            log::debug!("run_hooks args: {:?}", args);
+
+            let mut hook_command = process::Command::new(&hook.path)
+                .args(&args)
+                .arg0(&arg0)
+                .env_clear()
+                .envs(envs)
+                .stdin(process::Stdio::piped())
+                .spawn()
+                .with_context(|| "Failed to execute hook")?;
+            let hook_command_pid = Pid::from_raw(hook_command.id() as i32);
+            // Based on the OCI spec, we need to pipe the container state into
+            // the hook command through stdin.
+            if hook_command.stdin.is_some() {
+                let stdin = hook_command.stdin.take().unwrap();
+                serde_json::to_writer(stdin, state)?;
+            }
+
+            let res = if let Some(timeout_sec) = hook.timeout {
+                // Rust does not make it easy to handle executing a command and
+                // timeout. Here we decided to wait for the command in a
+                // different thread, so the main thread is not blocked. We use a
+                // channel shared between main thread and the wait thread, since
+                // the channel has timeout functions out of the box. Rust won't
+                // let us copy the Command structure, so we can't share it
+                // between the wait thread and main thread. Therefore, we will
+                // use pid to identify the process and send a kill signal. This
+                // is what the Command.kill() does under the hood anyway. When
+                // timeout, we have to kill the process and clean up properly.
+                let (s, r) = crossbeam_channel::unbounded();
+                thread::spawn(move || {
+                    let res = hook_command.wait();
+                    let _ = s.send(res);
+                });
+                match r.recv_timeout(time::Duration::from_secs(timeout_sec as u64)) {
+                    Ok(res) => res,
+                    Err(crossbeam_channel::RecvTimeoutError::Timeout) => {
+                        // Kill the process. There is no need to further clean
+                        // up because we will be error out.
+                        let _ = signal::kill(hook_command_pid, signal::Signal::SIGKILL);
+                        return Err(HookTimeoutError.into());
+                    }
+                    Err(_) => {
+                        unreachable!();
+                    }
+                }
+            } else {
+                hook_command.wait()
+            };
```

Comment:
I've refactored it a bit. What do you think?
```suggestion
            let mut hook_command = process::Command::new(&hook.path);
            if let Some((arg0, args)) = hook.args.as_ref().map(|a| a.split_first()).flatten() {
                log::debug!("run_hooks arg0: {:?}, args: {:?}", arg0, args);
                hook_command.arg0(arg0).args(args)
            } else {
                hook_command.arg0(&hook.path.as_path().display().to_string())
            };

            let mut hook_process = hook_command
                .env_clear()
                .envs(envs)
                .stdin(process::Stdio::piped())
                .spawn()
                .with_context(|| "Failed to execute hook")?;
            let hook_process_pid = Pid::from_raw(hook_process.id() as i32);
            // Based on the OCI spec, we need to pipe the container state into
            // the hook command through stdin.
            if let Some(stdin) = hook_process.stdin.as_ref() {
                serde_json::to_writer(stdin, state)?;
            }

            let res = if let Some(timeout_sec) = hook.timeout {
                // Rust does not make it easy to handle executing a command and
                // timeout. Here we decided to wait for the command in a
                // different thread, so the main thread is not blocked. We use a
                // channel shared between main thread and the wait thread, since
                // the channel has timeout functions out of the box. Rust won't
                // let us copy the Command structure, so we can't share it
                // between the wait thread and main thread. Therefore, we will
                // use pid to identify the process and send a kill signal. This
                // is what the Command.kill() does under the hood anyway. When
                // timeout, we have to kill the process and clean up properly.
                let (s, r) = crossbeam_channel::unbounded();
                thread::spawn(move || {
                    let res = hook_process.wait();
                    let _ = s.send(res);
                });
                match r.recv_timeout(time::Duration::from_secs(timeout_sec as u64)) {
                    Ok(res) => res,
                    Err(crossbeam_channel::RecvTimeoutError::Timeout) => {
                        // Kill the process. There is no need to further clean
                        // up because we will be error out.
                        let _ = signal::kill(hook_process_pid, signal::Signal::SIGKILL);
                        return Err(HookTimeoutError.into());
                    }
                    Err(_) => {
                        unreachable!();
                    }
                }
            } else {
                hook_process.wait()
            };
```

Reply:
I like this.